model/histogram/float_histogram.go

```diff
@@ -115,6 +115,295 @@ func (h *FloatHistogram) ZeroBucket() FloatBucket {
+// components of the receiving histogram. Buckets in the other histogram that do
+// not exist in the receiving histogram are inserted into the latter. The
+// resulting histogram might have buckets with a population of zero or directly
+// adjacent spans (offset=0). To normalize those, call the Compact method.
+//
+// This method returns a pointer to the receiving histogram for convenience.
+//
+// IMPORTANT: This method requires the Schema and the ZeroThreshold to be the
+// same in both histograms. Otherwise, its behavior is undefined.
+// TODO(beorn7): Change that!
+func (h *FloatHistogram) Add(other *FloatHistogram) *FloatHistogram {
+	h.ZeroCount += other.ZeroCount
+	h.Count += other.Count
+	h.Sum += other.Sum
+
+	// TODO(beorn7): If needed, this can be optimized by inspecting the
+	// spans in other and create missing buckets in h in batches.
+	iSpan, iBucket := -1, -1
+	var iInSpan, index int32
+	for it := other.PositiveBucketIterator(); it.Next(); {
+		b := it.At()
+		h.PositiveSpans, h.PositiveBuckets, iSpan, iBucket, iInSpan = addBucket(
+			b, h.PositiveSpans, h.PositiveBuckets, iSpan, iBucket, iInSpan, index,
+		)
+		index = b.Index
+	}
+	iSpan, iBucket = -1, -1
+	for it := other.NegativeBucketIterator(); it.Next(); {
+		b := it.At()
+		h.NegativeSpans, h.NegativeBuckets, iSpan, iBucket, iInSpan = addBucket(
+			b, h.NegativeSpans, h.NegativeBuckets, iSpan, iBucket, iInSpan, index,
+		)
+		index = b.Index
+	}
+	return h
+}
+
+// Sub works like Add but subtracts the other histogram.
+//
+// IMPORTANT: This method requires the Schema and the ZeroThreshold to be the
+// same in both histograms. Otherwise, its behavior is undefined.
+// TODO(beorn7): Change that!
+func (h *FloatHistogram) Sub(other *FloatHistogram) *FloatHistogram {
+	h.ZeroCount -= other.ZeroCount
+	h.Count -= other.Count
+	h.Sum -= other.Sum
+
+	// TODO(beorn7): If needed, this can be optimized by inspecting the
+	// spans in other and create missing buckets in h in batches.
+	iSpan, iBucket := -1, -1
+	var iInSpan, index int32
+	for it := other.PositiveBucketIterator(); it.Next(); {
+		b := it.At()
+		b.Count *= -1
+		h.PositiveSpans, h.PositiveBuckets, iSpan, iBucket, iInSpan = addBucket(
+			b, h.PositiveSpans, h.PositiveBuckets, iSpan, iBucket, iInSpan, index,
+		)
+		index = b.Index
+	}
+	iSpan, iBucket = -1, -1
+	for it := other.NegativeBucketIterator(); it.Next(); {
+		b := it.At()
+		b.Count *= -1
+		h.NegativeSpans, h.NegativeBuckets, iSpan, iBucket, iInSpan = addBucket(
+			b, h.NegativeSpans, h.NegativeBuckets, iSpan, iBucket, iInSpan, index,
+		)
+		index = b.Index
+	}
+	return h
+}
+
+// addBucket takes the "coordinates" of the last bucket that was handled and
+// adds the provided bucket after it. If a corresponding bucket exists, the
+// count is added. If not, the bucket is inserted. The updated slices and the
+// coordinates of the inserted or added-to bucket are returned.
+func addBucket(
```

Comment:
I am finding it little tricky to follow this code. So I will try the other approach that I mentioned in another comment and try to replace this with that